docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+changes. Every time you modify an automerge document you generate a new entry 
+for the change log. Each change is just some bytes. When you receive changes 
+from other actors you just pass these changes (which, again, are just bytes) to
+automerge to add to the change log. The end result is that you load a bunch of
+binary changes and get back a JSON object.
+
+There are some subtleties around preserving user intent when modifying
+documents, but these are not too onerous.
+
+### Object Trees
+
+Given that automerge changes are a hash linked graph, we can map them to Git.
+We do so by wrapping each change in a commit. The commit points at a tree with
+the following layout
+
+```
+.
+|--manifest
+|--schema
+|  |--schema.json
+|  |--migrations
+|  |  |--<migration hash>.json
+|  |  |--<migration hash>.json
+|--change
+|--author
+```
+
+This tree contains a single change to a collaborative object. We will go into
+more details shortly. Any direct dependencies of this change are encoded in the 
+same manner and become the parents of this commit. This allows us to
+reconstruct the automerge depdency graph. 
+
+Along with the dependencies of the commit we also need to add the commit of the
+identity which created this commit. We need this identity to validate
+signatures and by making the commit a parent we ensure that git will replicate
+it for us.
+
+
+#### `manifest`
+
+The manifest is a TOML file containing some metadata about the object.
+Specifically it will contain:
+
+- `id`, a UUID, generated at the time the object is created
+- `typename`, discussed above
+- `history_type`, always `"automerge"`, this is here to allow for different
+  CRDT implementations in future.
+
+
+Each object is also created with a JSON schema. The schema is represented by an
+initial `schema.json` and a series of schema migrations which extend that
+initial schema. Schema migrations will not be addressed in detail in this RFC
+but we will show their feasibility.
+
+#### `schema`
+
+Schemas are primarily important for the interoperability of the system. We need
+applications to be able to rely on the data they are working with being valid,
+otherwise we impose the problem of schema validation on application developers.
+We represent schemas using JSON schema which is in the `schema/schema.json` of any
+object tree.
+
+Schemas will need to be able to change, the `schema/migrations` directory is
+present to allow us to store compatible changes to a schema in future. Schema
+migration is out of scope for this RFC.
+
+
+#### `change`
+
+This is the automerge change which this commit introduces. It is a binary file
+which must contain a single change and it's dependents must be the dependents
+referenced by the parents of the commit.
+
+
+#### `author`
+
+This is a file containing the SHA of the commit which references the authors
+identity. We need this so we know to ignore this commit when walking the
+history to look for changes.
```

Comment:
I'm not sure I understand this, could you expand on it?

Reply:
The purpose of representing each change as a hash is to encode the dependency graph of the automerge document into Git. This has a few nice properties:

- We know that if we have `fetch`ed a change then we have also fetched all of the dependencies of that change
- We can deduplicate storage of common dependencies, even across remotes. This is useful if many people are (for example) responding to an issue, they can all depend on each others changes without copying them on disk.

However, when we're validating the hash graph we also need to ensure that the identity of the author of the change is present so that we can check the signature on the change. Therefore we add the commit of the head of the author identity as a parent of the change commit to ensure that it is fetched along with the change. 

Now, we want to be able to reconstruct the automerge change graph from the git change graph. We do this by starting at the latest change and walking the commit tree, recursively adding each parent to the automerge change graph. As we do this we want to exclude the parent commits which refer to the signing identity rather than an automerge change, but there's no way to tell whether a commit is an automerge change or not without examining it, so the `author` file contains the hash of the commit which is the identity so we can exclude it from our tree walk as we construct the automerge change graph.

Reply:
Ah that explanation was super helpful!

Reply:
Is the hash of a `change` blob actually guaranteed to be stable?

There is also the opportunity to not having to follow the blob pointer of `author` by either encoding the SHA1 in the file name itself (like git notes), or adding a trailer to the commit (a la `x-rad-author:`).